Automattic/woocommerce-payments
# Implement a retry system to the WCPay Subscription migration process

Fixes #7038

#### Changes proposed in this Pull Request









When an exception is thrown during our migration process, the error is logged and the subscription remains as a Stripe Billing subscription until the merchant manually runs the migration script again.

In this PR we're implementing a retry system so that if the migration fails for whatever reason, another attempt is automatically scheduled in the future. With each failed attempt (up to 7 retries), we will increase the time between retrying the migration again.

Current retry schedule:
1. after 1 minute
2. after 5 minutes
3. after10 minutes
4. after 30 minutes
5. after 1 hour
6. after 6 hours
7. after 12 hours

When does a migration get rescheduled?

- When its number of attempts is less than 7
- When the `migrate_wcpay_subscription()` function throws a non-skip error/exception
- When the scheduled action throws an unexpected failure
- When an unexpected shutdown occurs

This PR also includes a few changes that aren't directly related to the retry system:
 - 4e6ebc73ceea25640164c5c9351c409aae7dd6d8 Manually hook onto the `wcpay_migrate_subscription` hook so that we can use two args on the hook
 - 3b69426b7dc23170ac45898e9b82325e5f28dc3f Update some of the logged messages to always include a subscription ID and/or WCPay subscription ID so that we have the context in the single log.

#### Testing instructions





I wanted to test some failing migrations and some successful ones in the same batch so I injected some faulty code surrounded by rand(0,1) to simulate errors being caused 50% of the time.

Here are some examples of code I used:

To throw unexpected errors:

```php
if ( rand( 0, 1 ) ) {
	throw new Error( 'foobar' );
}
```
```php
if ( rand( 0, 1 ) ) {
 $this->api_client = null;
}
```

To trigger Stripe API errors I updated the [`cancel_subscription()`](https://github.com/Automattic/woocommerce-payments/blob/e6e7c3be3d06153233bd14159bb2c4f0f1e120b4/includes/wc-payment-api/class-wc-payments-api-client.php#L1181) and [`get_subscription()`](https://github.com/Automattic/woocommerce-payments/blob/e6e7c3be3d06153233bd14159bb2c4f0f1e120b4/includes/wc-payment-api/class-wc-payments-api-client.php#L1129) functions to sometimes fetch subscriptions that don't exist:
```php
public function cancel_subscription( string $wcpay_subscription_id ) {
	return $this->request(
		[],
		self::SUBSCRIPTIONS_API . '/' . $wcpay_subscription_id . ( rand( 0, 1 ) ? '1234' : '' ),
		self::DELETE
	);
}
```

To test that skipped migrations don't get retried I manually added a scheduled action for a subscription I just migrated using WP Console:

```php
as_schedule_single_action( gmdate( 'U' ), 'wcpay_migrate_subscription', [ 'migrate_subscription' => 1918 ] );
```

**Testing instructions:**

1. Generate a bunch of WCPay Subscriptions using [this plugin/gist](https://gist.github.com/james-allan/38fa3f40fff586a8377bf629f380342c).
2. Enable migrations by setting `_wcpay_feature_allow_subscription_migrations` to `1` in `wp-options`
3. Enable WC Subscriptions extension
4. Go to **WooCommerce > Status > Tools** and manually run the migration tool
5. Go to **WooCommerce > Status > Logs** and search look in the latest `woopayments-subscription-migration`
6. Confirm logs are showing migrations being rescheduled on failures and the number of attempts are in the logs.
7. Open Action Scheduler and search `wcpay_migrate_subscription` to confirm that some migrations have an `attempt` arg.

In order to test the retry schedule and migration attempts stop after the 7th retry, I removed the rand(0,1) and made it so an error is always thrown. Then manually ran the migration action.

-------------------


**Post merge**
